### PR TITLE
Add "set requester secret" endpoint

### PR DESCRIPTION
**Pull request by mcornaton**

### CODE REVIEW

File: src/main/java/com/iexec/sms/authorization/AuthorizationService.java [hunk 1]
```diff
@@ -149,6 +149,18 @@ public String getChallengeForSetAppRequesterRuntimeSecretCount(
                 Long.toHexString(secretCount));
     }
 
+    public String getChallengeForSetAppRequesterRuntimeSecret(
```
Comment (jeremyjams) [inline src/main/java/com/iexec/sms/authorization/AuthorizationService.java:152]:
```suggestion
    public String getChallengeForSetAppRequesterAppRuntimeSecret(
```

File: src/main/java/com/iexec/sms/secret/app/AppRuntimeSecretController.java [hunk 2]
```diff
@@ -80,7 +80,7 @@ public AppRuntimeSecretController(AuthorizationService authorizationService,
         return ResponseEntity.noContent().build();
     }
 
-    @RequestMapping(method = RequestMethod.HEAD, path = "/{appAddress}/secrets/{secretIndex}")
+    @RequestMapping(method = RequestMethod.HEAD, path = "/apps/{appAddress}/secrets/{secretIndex}")
     public ResponseEntity<Void> isAppRuntimeSecretPresent(@PathVariable String appAddress,
```
Comment (jeremyjams) [inline src/main/java/com/iexec/sms/secret/app/AppRuntimeSecretController.java:84]:
```suggestion
    public ResponseEntity<Void> isAppDeveloperAppRuntimeSecretPresent(@PathVariable String appAddress,
```

File: src/main/java/com/iexec/sms/secret/app/AppRuntimeSecretController.java [hunk 3]
```diff
@@ -99,7 +99,7 @@ public AppRuntimeSecretController(AuthorizationService authorizationService,
         return ResponseEntity.notFound().build();
     }
 
-    @PostMapping("/{appAddress}/requesters/secrets")
+    @PostMapping("/apps/{appAddress}/requesters/secrets")
     public ResponseEntity<String> setAppRequestersAppRuntimeSecretCount(
```
Comment (jeremyjams) [inline src/main/java/com/iexec/sms/secret/app/AppRuntimeSecretController.java:103]:
```suggestion
    public ResponseEntity<String> setAppRequesterAppRuntimeSecretCount(
```

**Review by jeremyjams — COMMENTED** [review]
Very nice!

File: src/main/java/com/iexec/sms/secret/app/AppRuntimeSecretController.java [hunk 4]
```diff
@@ -142,4 +142,66 @@ public AppRuntimeSecretController(AuthorizationService authorizationService,
...
+                    requesterAddress, appAddress, secretIndex);
+            return ResponseEntity.badRequest().build();
+        }
+
+        if (secretIndex < 0) {
+            log.error("Negative index are forbidden for app requester secrets." +
+                            " [requesterAddress: {}, appAddress: {}, secretIndex: {}]",
+                    requesterAddress, appAddress, secretIndex);
+            return ResponseEntity.badRequest().build();
+        }
```
[the remark this answers is not in the document]
Comment (mcornaton) [reply src/main/java/com/iexec/sms/secret/app/AppRuntimeSecretController.java:169]: Yes, that's a mistake. That's now fixed :)

File: src/main/java/com/iexec/sms/secret/teetaskruntime/AppRuntimeSecretController.java [hunk 5]
```diff
@@ -147,6 +152,95 @@ public AppRuntimeSecretController(AuthorizationService authorizationService,
 
         return ResponseEntity.noContent().build();
     }
+    // endregion
+
+    // region App requester endpoint
+    @PostMapping("/requesters/{requesterAddress}/apps/{appAddress}/secrets/{secretIndex}")
+    public ResponseEntity<String> addAppRequesterAppRuntimeSecret(@RequestHeader("Authorization") String authorization,
```
Comment (jeremyjams) [inline src/main/java/com/iexec/sms/secret/teetaskruntime/AppRuntimeSecretController.java:159]: You can update role name everywhere to ~~AppRequester~~ Requester.

File: src/main/java/com/iexec/sms/secret/teetaskruntime/AppRuntimeSecretController.java [hunk 6]
```diff
@@ -147,6 +152,95 @@ public AppRuntimeSecretController(AuthorizationService authorizationService,
...
+                secretIndex)) {
+            log.error("Unauthorized to addAppRequesterAppRuntimeSecret: " +
+                            "secret already exists" +
+                            " [requesterAddress:{}, appAddress:{}, secretIndex:{}]",
+                    requesterAddress, appAddress, secretIndex);
+            return ResponseEntity.status(HttpStatus.CONFLICT).build(); // secret already exists
+        }
+
+        final Optional<TeeTaskRuntimeSecretCount> oAllowedSecretsCount =
+                teeTaskRuntimeSecretCountService.getAppRuntimeSecretCount(
```
Comment (jeremyjams) [inline src/main/java/com/iexec/sms/secret/teetaskruntime/AppRuntimeSecretController.java:211]:
```suggestion
                teeTaskRuntimeSecretCountService.getRequesterAppRuntimeSecretCount(
```

File: src/main/java/com/iexec/sms/secret/teetaskruntime/AppRuntimeSecretController.java (hunk 5 above)
Comment (mcornaton) [reply src/main/java/com/iexec/sms/secret/teetaskruntime/AppRuntimeSecretController.java:159]: Sure, I'll do it in the last PR: https://github.com/iExecBlockchainComputing/iexec-sms/pull/76

Comment (mcornaton) [reply src/main/java/com/iexec/sms/secret/teetaskruntime/AppRuntimeSecretController.java:159]: Done there: https://github.com/iExecBlockchainComputing/iexec-sms/commit/a6e17fd41d18dda94b9133dbb34ecbbb0696c336

File: src/main/java/com/iexec/sms/secret/teetaskruntime/AppRuntimeSecretController.java (hunk 6 above)
Comment (mcornaton) [reply src/main/java/com/iexec/sms/secret/teetaskruntime/AppRuntimeSecretController.java:211]: We pass the secret owner role so I don't think the method name should include `Requester`.

Comment (jeremyjams) [reply src/main/java/com/iexec/sms/secret/teetaskruntime/AppRuntimeSecretController.java:211]: Ok I see so we can discard my comment.
But we should rename it to something like getMaxAppRuntimeSecretCount

Comment (mcornaton) [reply src/main/java/com/iexec/sms/secret/teetaskruntime/AppRuntimeSecretController.java:211]: Sure, that looks great! Thanks :)

Comment (mcornaton) [reply src/main/java/com/iexec/sms/secret/teetaskruntime/AppRuntimeSecretController.java:233]: Done. I have renamed all "secret count" method into "max secret count"

**Pull request merged**